icu4c/source/test/cintltst/cbiapts.c

```diff
@@ -515,6 +517,79 @@ static void TestBreakIteratorSafeClone(void)
         ubrk_close(someIterators[i]);
     }
 }
+
+static void TestBreakIteratorClone(void)
+{
+    UChar text[51];     /* Keep this odd to test for 64-bit memory alignment */
+                        /*  NOTE:  This doesn't reliably force mis-alignment of following items. */
+    UBreakIterator * someIterators [CLONETEST_ITERATOR_COUNT];
+    UBreakIterator * someClonedIterators [CLONETEST_ITERATOR_COUNT];
+
+    UErrorCode status = U_ZERO_ERROR;
+    int32_t start,pos;
+    int32_t i;
+
+    /*Testing ubrk_clone */
+
+    /* Note:  the adjacent "" are concatenating strings, not adding a \" to the
+       string, which is probably what whoever wrote this intended.  Don't fix,
+       because it would throw off the hard coded break positions in the following
+       tests. */
+    u_uastrcpy(text, "He's from Africa. ""Mr. Livingston, I presume?"" Yeah");
+
+    /* US & Thai - rule-based & dictionary based */
+    someIterators[0] = ubrk_open(UBRK_WORD, "en_US", text, u_strlen(text), &status);
+    if(!someIterators[0] || U_FAILURE(status)) {
+      log_data_err("Couldn't open en_US word break iterator - %s\n", u_errorName(status));
+      return;
+    }
+
+    someIterators[1] = ubrk_open(UBRK_WORD, "th_TH", text, u_strlen(text), &status);
+    if(!someIterators[1] || U_FAILURE(status)) {
+      log_data_err("Couldn't open th_TH word break iterator - %s\n", u_errorName(status));
+      return;
+    }
+
+    /* test each type of iterator */
+    for (i = 0; i < CLONETEST_ITERATOR_COUNT; i++)
+    {
+
+        /* Check the various error & informational states */
+
+        /* Null status - just returns NULL */
+        if (NULL != ubrk_clone(someIterators[i],  NULL))
+        {
+            log_err("FAIL: Cloned Iterator failed to deal correctly with null status\n");
+        }
+        /* error status - should return 0 & keep error the same */
+        status = U_MEMORY_ALLOCATION_ERROR;
+        if (NULL != ubrk_clone(someIterators[i], &status) || status != U_MEMORY_ALLOCATION_ERROR)
+        {
+            log_err("FAIL: Cloned Iterator failed to deal correctly with incoming error status\n");
+        }
+
+        status = U_ZERO_ERROR;
+        /* Null Iterator - return NULL & set U_ILLEGAL_ARGUMENT_ERROR */
+        if (NULL != ubrk_clone(NULL, &status) || status != U_ILLEGAL_ARGUMENT_ERROR)
+        {
+            log_err("FAIL: Cloned Iterator failed to deal correctly with null Iterator pointer\n");
+        }
+        status = U_ZERO_ERROR;
+
+        /* Do these cloned Iterators work at all - make a first & next call */
+        someClonedIterators[i] = ubrk_clone(someIterators[i], &status);
+
+        start = ubrk_first(someClonedIterators[i]);
+        if(start!=0)
+            log_err("error ubrk_start(clone) did not return 0\n");
+        pos=ubrk_next(someClonedIterators[i]);
+        if(pos!=4)
+            log_err("error ubrk_next(clone) did not return 4\n");
+
+        ubrk_close(someClonedIterators[i]);
```

Comment:
so... if we clone someClonedIterators[i] from here and call ubrk_next() , the 2nd clone copy should clone the state and make sure the next ubrk_next() call not start from 0 but from 4, right? Could we add test to ensure the clone not only clone the text but also other states?

Reply:
FYI -- I have not verified this -- but often we only have C API coverage testing in cintltst (pure C), and more interesting functional testing happens in C++ intltest. Gives us coverage on the thin C wrapper code, but implementation coverage happens in C++. I assume/hope that Andy did decent functional testing in C++.

Reply:
Added the test for keeping the text position after the clone operations. Please take a look.